File: back-end/crates/malbox-cli-common/src/api/images.rs
use super::ApiClient;
use crate::error::Result;
use serde::{Deserialize, Serialize};

#[derive(Serialize)]
pub struct RegisterImageRequest {
    pub name: String,
    pub platform: String,
    pub arch: String,
    pub format: Option<String>,
    pub description: Option<String>,
    pub path: String,
}

#[derive(Deserialize, Debug)]
pub struct Image {
    pub id: Option<serde_json::Value>,
    pub name: String,
    pub platform: Option<serde_json::Value>,
    pub arch: Option<serde_json::Value>,
    pub format: Option<String>,
    pub description: Option<String>,
    pub path: Option<String>,
    pub available: Option<bool>,
    pub created_at: Option<serde_json::Value>,
    pub updated_at: Option<serde_json::Value>,
}

impl ApiClient {
    pub async fn register_image(&self, request: RegisterImageRequest) -> Result<Image> {
        let response = self
            .client
            .post(self.url("/v1/images"))
            .json(&request)
            .send()
            .await?;
        let response = self.check_response(response).await?;
        Ok(response.json().await?)
    }

    pub async fn list_images(&self) -> Result<Vec<Image>> {
        let response = self.client.get(self.url("/v1/images")).send().await?;
        let response = self.check_response(response).await?;
        Ok(response.json().await?)
    }

    pub async fn get_image(&self, name: &str) -> Result<Image> {
        let response = self
            .client
            .get(self.url(&format!("/v1/images/{}", name)))
            .send()
            .await?;
        let response = self.check_response(response).await?;
        Ok(response.json().await?)
    }

    pub async fn delete_image(&self, name: &str) -> Result<()> {
        let response = self
            .client
            .delete(self.url(&format!("/v1/images/{}", name)))
            .send()
            .await?;
        self.check_response(response).await?;
        Ok(())
    }
}

File: front-end/src/lib/api/images.ts
import { request, requestJson, type FetchLike } from './client';
import type { Image, RegisterImageRequest } from './types';

export async function listImages(fetchFn: FetchLike): Promise<Image[]> {
	return requestJson<Image[]>(fetchFn, '/v1/images');
}

export async function getImage(fetchFn: FetchLike, name: string): Promise<Image> {
	return requestJson<Image>(fetchFn, `/v1/images/${encodeURIComponent(name)}`);
}

export async function registerImage(
	fetchFn: FetchLike,
	req: RegisterImageRequest
): Promise<{ id: string; name: string; path: string; available: boolean }> {
	return requestJson(fetchFn, '/v1/images', {
		method: 'POST',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify(req)
	});
}

export async function deleteImage(fetchFn: FetchLike, name: string): Promise<void> {
	await request(fetchFn, `/v1/images/${encodeURIComponent(name)}`, { method: 'DELETE' });
}
